From: Mel Gorman <mgorman@suse.de>
Date: Fri, 13 Sep 2013 10:25:31 +0100
Subject: [PATCH] mm: vmscan: take page buffers dirty and locked state into
 account

References: Limit reclaim in the preserve of IO (bnc#754690)
Patch-mainline: No, never

There is no guarantee that ext3 filesystem is in use and the writepage
aops can be easily NULL.

Signed-off-by: Mel Gorman <mgorman@suse.de>
---
 mm/vmscan.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index ec8ce6d..abaf914 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -708,7 +708,8 @@ static void page_check_dirty_writeback(struct page *page,
 		return;
 
 	mapping = page_mapping(page);
-	if (mapping && (mapping->a_ops->writepage == ext3_aops_writepage ||
+	if (mapping && mapping->a_ops->writepage &&
+			(mapping->a_ops->writepage == ext3_aops_writepage ||
 			 mapping->a_ops->writepage == blkdev_aops_writepage))
 		buffer_check_dirty_writeback(page, dirty, writeback);
 }
